README.md
Python Data Streamer
===============

[![Join the chat at https://gitter.im/InitialState/python_appender](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/InitialState/python_appender?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) 

[![Build Status](https://travis-ci.org/initialstate/python_appender.svg?branch=master)](https://travis-ci.org/initialstate/python_appender)

This is a Python module currently built for python >= 2.7

##Installation
###Using the automated script

On a Unix based system: (including Raspberry Pi, Mac OS X, Ubuntu) 

```
\curl -sSL https://get.initialstate.com/python -o - | sudo bash
```

If you don't have `curl` you should get it

```
sudo apt-get install curl
```

```
sudo yum install curl
```


###Using Package Management
The package is hosted in PyPI under the package name [ISStreamer](https://pypi.python.org/pypi/ISStreamer).

####If you don't have `pip`:

1. (*optional*) Check if you have python setup tools installed:

	```
	$ easy_install --version
	```
	
	if you don't see a version number come back, you should install setuptools:
	
	```
	$ sudo apt-get install python-setuptools
	```
	

2. (*optional*) Check if you have pip installed:

	```
	$ pip --version
	```
	
	if you don't see a version number after running the above command, install pip using easy_install:
	
	```
	$ sudo easy_install pip
	```

####I've got `pip` what next?:


```
$ sudo pip install ISStreamer
```

> This command installs the ISStreamer module
	

##Basic Usage

After getting the ISStreamer module, usage is really simple. With the following example, you can do most of what you need, and you **don't need to read further**! However


```python
from ISStreamer.Streamer import Streamer

# create a Streamer instance
streamer = Streamer(bucket_name="Some Bucket Name", bucket_key="bucket_key", access_key="YourAccessKey")

# send some data
streamer.log("test", "hi")
streamer.log("temperature", 32)

# flush and close the stream
streamer.close()
```


##Advanced Usage, Troubleshooting, and Concepts


###Concepts
- ####Buckets

	In order to keep your event streams and visualizations contextually appropriate, we have implemented a concept called `buckets`. A new bucket is automatically created when the Streamer is constructed, however, if you want to append to an existing bucket, or use a key that is more memorable than the uuid that will otherwise be used, you can use the optional `bucket_key` constructor parameter. If a Streamer is constructed with a `bucket_key` that already exists, then any data sent in that Stream will append to the existing bucket.

	> NOTE: `bucket_key`'s uniqueness is scoped to a specific `access_key`.

- ####Event Stream

	An event stream is a key with an associated set of values with timestamps. These individual events are created every time the `Strmeaer.log` method is called. If an `event_key` is the same for different pieces of data, those pieces of data are represented together in an event stream. This is more of an Initial State concept than a Python Streamer specific concept.

	> Legacy Note: Event Streams use to be called Signals.

###Most Used Methods
- ####`Streamer.log(key, value[, epoch])`
	This is the core method and api for the event streamer. This is an asyncronous method that pushes your data to a queue that handles sending it off to Initial State's servers. You don't have to worry about anything but calling the method where you want! For the sake of clarity (for those new to python or programming) the `Streamer` would be replaced with the variable reference to a `Streamer` instance.
	
	The `log` method expects two parameters, `key` and `value`:
	- `key` is a string that represents the source of the `value` and is limited to 250 characters
	- `value` is either a string, boolean, or number and it represents a value at the time of the method call.
	- `epoch` is an optional parameter to override the epoch timestamp, recommended for advanced uses.

- ####`Streamer.log_object(obj[, key_prefix[, epoch]])`
	This is an enhanced method to abstract having to write a bunch of log statements to stream all of the values of an object with multiple data points at a specific time.

	The `log_object` method expects one parameter, `obj`:
	- `obj` is either a list, dict, or simple object with attributes.
		- If `obj` is a list, it will use the key name `list_n` unless the optional `key_prefix` is supplied, then it will use `key_prefix_n` where `n` - in both cases - is an index.
		- If `obj` is a dict, it will use the key name `dict_key` where unless the optional `key_prefix` is supplied, then it will use `key_prefix_key` where `key` - in both cases - is the key of the dictionary value.
		- If `obj` is a simple object, it will iterate over the objects attributes and produce values for keys with the name of the key as `obj_attr` unless the `key_prefix` is supplied, then it will use `key_prefix_attr`. In all cases, `attr` is the attribute name.
	- `key_prefix` is an optional string that, when supplied, overrrides the default key prefixes.
	- `epoch` is an optional number that represents the current time in epoch format.

	[Here is a working example](/example_app/example_compute_metrics.py).

	> NOTE: log_object will log multiple keys and values, but will override the epoch timestamp of each value so that there is no cpu or iteration skew in the timestamp reported for when those values were logged and streamed.

- ####`Streamer.close()`
	This method ensures that the log buffer is flushed and should be called when a program is exiting. It is also called during the `__del__` magic method of the `Streamer` by python, but it is a best practice to explicitly call it at the end of a program to ensure it is executed.

###Advanced Use
- ####Manual `Streamer.flush()`
	You can manually flush on your own accord by calling `Streamer.flush()`. This will ensure that anything that has been queued or buffered locally  will get sent to Initial State's servers asap.
	
- ####Changing buffer size
	You can override the default event buffer size (the count of events) by passing the optional `buffer_size` parameter into the Streamer constructor. Here is an example:

	```python
	streamer = Streamer(bucket_name="Hi!", access_key="YourAccessKey", buffer_size=200)
	```

	In this example, the `buffer_size` is being increased to 200 from the default, 10. The decision to override this value should be based on how many log statements you make in a loop before sleeping. You can typically play around with this number to help tune the performance of the Streame. Additionally, a higher buffer_size may help if getting throttled by the api.

	Another great example of overriding the buffer_size would be to increase to a large value and use the `Streamer.flush()` method at the end of each iteration to effectively have a dynamic buffer size. Here is an example:

	```python
	...

	streamer = Streamer(bucket_name="Dynamic Buffer", access_key="YourAccessKey", buffer_size=200)

	counter = 0
	while 1:
		streamer.log("loop_counter", counter)

		some_dict = {
			"a": 1,
			"b": 2,
			"c": 3
		}
		streamer.log_object(some_dict, key_prefix="some_dict")

		dynamic_list = SomeOtherModule.PracticalClass.get_stuff()
		
		# We don't know how many items will be in dynamic_list, so we just
		# log the list to get them all. However, since we don't know how many
		# there will be, we don't know the optimal buffer size, so we set the
		# buffer to a high value and flush at the end of the iteration
		streamer.log_object(dynamic_list)

		# Here we will flush the log buffer to insure that there isn't a delay
		# in sending the logs that we've created previous to this point while
		# waiting for the sleep to finish
		streamer.flush()

		# increment counter
		counter += 1

		# sleep for 10 seconds
		time.sleep(10)

	...
	```

- ####Overriding the timestamp
	Some have asked for the ability to override the timestamp. Currently, the timestamp is automatically associated with data by retrieving the most accurate timestamp possible from the device as soon as a `log` or `log_object` method is called. However, you can override this by doing the following:

	```python

	time = time.time()

	streamer.log("siganl", 5, epoch=time)

	```

	For a full example checkout [this](/example_app/time_override_example.py)


- ####Creating a new bucket
	When you construct a `Streamer` the constructor expects a name or a key that it will use to ensure there is a bucket that it will use as the context for `Streamer.log(key, value)`. Buckets are either created or consumed based on the unique combination of a `access_key` and a `bucket_key`. If you want to switch which to a new bucket, because say you've started a new session or run, simply call `Streamer.set_bucket(bucket_name='some_bucket_name'[, bucket_key='some_bucket_key'])`. Note that bucket_key is optional, if not provided the module will create a `uuid4` as the `bucket_key`. Here is an example:
	
	```python
	streamer = Streamer(bucket_name="Starting Bucket", access_key="YourAccessKey")
	
	streamer.log("key1", "starting")
	streamer.set_bucket(bucket_name="New Bucket")
	streamer.log("key1", "starting")
	```  

	In this example, you will get a key1=starting in two different buckets: "Starting Bucket" and "New Bucket".


###Troubleshooting
####Missing Events
If the Streamer cannot ship a set of events during a flush, it will retry a few times before deeming it a failure. If it does fail, it will attempt to save it's payload to a local file. This payload will be in a json format inside a json array. Each array can be individually submitted to Initial State's events api to fill in any missed events.

####Setting `debug_level`
If you're having issues with your data you might want to try running ISStreamer at a higher debug level:

```python
logger = Streamer(bucket_name="SomeBucketName", access_key="YourAccessKey", debug_level=2)
```

With a `debug_level` at or greater than 2 the streamer will throw exceptions on logging errors. Otherwise, it will assume logging errors are not fundamentally exceptional. It will also display more verbouse logging information.

> If you want to simply see what's being logged in the console, construct the Streamer object with the debug_level equal to 1.


CHANGES.txt
0.2.0 (2017-03-31)
-------------------

* exposing rate limiting
* added `async` flag to `Streamer` constructor. When set to `async=false` Streamer will block when shipping to Initial State API in order to wait for the response. 
* added a wait time based on the `Retry-After` header responded by the API when a `429` response is received 
* using Initial State API >=0.0.4
* added the ability to set the key_prefix="" on log_object so that no prefix of `_` character are added to the key of the object stream

0.0.26
-------------------

* bubbling up exception for 402 responses from api to be clearer.
* added a missed event log so that missed events can be put into a file instead of to std out
* FIXED: logic in 402 responses


0.0.25 (2015-02-12)
-------------------

* Entirely new to support new initial state streaming api
* api is versioned
* only one api endpiont for streaming events and creating buckets
* BREAKING: new syntax for constructor parameters: `bucket` -> `bucket_name`, `client_key` -> `access_key`
* BREAKING: `log_object` parameters: `signal_prefix` -> `key_prefix`
* BREAKING: `sessionId` is now `bucket_key` which can be configured in the Streamer constructor.


0.0.24 (2014-11-24)
-------------------

- Nothing changed yet.


0.0.23 (2014-11-21)
-------------------

- Nothing changed yet.


0.0.22 (2014-11-21)

* feature: added log_object method to allow automatically logging flat objects attributes, dicts, and lists
* feature: improved example_compute_metrics.py
* feature: added object_log_example.py
* feature: install script added check for pip version for `--pre` flag. added ability to specify wether or not to get pre-released streamer
* fix: added condition in install script to ignore using `--pre` flag when pip version is less than 1.4
* feature: added ability to override epoch timestamp via optional `epoch` parameter on `Streamer.log(signal, value[, epoch])`
* feature: added ability to override buffer size before shipping logs via optional constructor parameter `Streamer([buffer_size])`
* feature: added CHANGES.txt to repository to start documenting changes
* fix: added more status code checks to bucket creation to return more descriptive message for incorrect `client_key`

ISStreamer/Streamer.py
# local config helper stuff
try:
	import ISStreamer.configutil as configutil
except ImportError:
	import configutil
try:
	import ISStreamer.version as version
except ImportError:
	import version
import uuid

# python 2 and 3 conversion support
import sys
if (sys.version_info < (2,7,0)):
	sys.stderr.write("You need at least python 2.7.0 to use the ISStreamer")
	exit(1)
elif (sys.version_info >= (3,0)):
	import http.client as httplib
else:
	import httplib
import json
# time stuff
import datetime
import time
# performance stuff
import threading
import collections
import csv

class Streamer:
	BucketName = ""
	AccessKey = ""
	Channel = ""
	BufferSize = 10
	StreamApiBase = ""
	LogQueue = None
	DebugLevel = 0
	BucketKey = ""
	IsClosed = True
	Offline = False
	Async = True
	LocalFile = None
	ApiVersion = '<=0.0.4'
	MissedEvents = None
	def __init__(self, bucket_name="", bucket_key="", access_key="", ini_file_location=None, debug_level=0, buffer_size=10, offline=None, async=True):
		config = configutil.getConfig(ini_file_location)
		if (offline != None):
			self.Offline = offline
		else:
			if (config["offline_mode"] == "false"):
				self.Offline = False
			else:
				self.Offline = True

		self.Async = async
		if (self.Offline):
			try:
				file_location = "{}.csv".format(config["offline_file"])
				self.LocalFileHandler = open(file_location, 'w')
				self.LocalFile = csv.writer(self.LocalFileHandler)
			except:
				print("There was an issue opening the file (nees more description)")

		if (config == None and bucket_name=="" and access_key == ""):
			raise Exception("config not found and arguments empty")

		if (bucket_name == ""):
			bucket_name = config["bucket"]
		else:
			bucket_name = bucket_name
		if (access_key == ""):
			self.AccessKey = config["access_key"]
		else:
			self.AccessKey = access_key


		#self.LogQueue = Queue.Queue(self.BufferSize)
		self.BucketKey = bucket_key
		self.BufferSize = buffer_size
		self.LogQueue = collections.deque()

		self.StreamApiBase = config["stream_api_base"]
		self.set_bucket(bucket_name, bucket_key)
		self.DebugLevel = debug_level
		self.IsClosed = False

		self.console_message("access_key: {accessKey}".format(accessKey=self.AccessKey))
		self.console_message("stream_api_base: {api}".format(api=self.StreamApiBase))
	
	def ship_to_api(self, resource, contents):
		api_base = self.StreamApiBase
		
		headers = {
			'Content-Type': 'application/json',
			'User-Agent': 'PyStreamer v' + version.__version__,
			'Accept-Version': self.ApiVersion,
			'X-IS-AccessKey': self.AccessKey,
			'X-IS-BucketKey': self.BucketKey
		}

		def __ship(retry_attempts, wait=0):
			conn = None
			response = None
			if (self.StreamApiBase.startswith('https://')):
				api_base = self.StreamApiBase[8:]
				self.console_message("ship {resource}: stream api base domain: {domain}".format(domain=api_base, resource=resource), level=2)
				conn = httplib.HTTPSConnection(api_base, timeout=120)
			else:
				api_base = self.StreamApiBase[7:]
				self.console_message("ship {resource}: stream api base domain: {domain}".format(domain=api_base, resource=resource), level=2)
				conn = httplib.HTTPConnection(api_base, timeout=120)

			retry_attempts = retry_attempts - 1
			if (retry_attempts < 0):
				if (self.DebugLevel >= 2):
					raise Exception("shipping failed.. network issue?")
				else:
					self.console_message("ship: ISStreamer failed to ship after a number of attempts.", level=0)
					if (self.MissedEvents == None):
						self.MissedEvents = open("err_missed_events.txt", 'w+')
					if (self.MissedEvents != None):
						self.MissedEvents.write("{}\n".format(json.dumps(contents)))
					return
			
			try:
				if (wait > 0):
					self.console_message("ship-debug: pausing thread for {wait} seconds".format(wait=wait))
					time.sleep(wait)

				conn.request('POST', resource, json.dumps(contents), headers)
				response = conn.getresponse()

				if (response.status >= 200 and response.status < 300):
					self.console_message("ship: status: " + str(response.status) + "\nheaders: " + str(response.msg), level=2)
					self.console_message("ship: body: " + str(response.read()), level=3)
				elif (response.status == 401 or response.status == 403):
					self.console_message("ERROR: unauthorized access_key: " + self.AccessKey)
				elif (response.status == 402):
					self.console_message("AccessKey exceeded limit for month, check account at https://app.initialstate.com/#/account")
					raise Exception("PAYMENT_REQUIRED")
				elif (response.status == 429):
					if "Retry-After" in response.msg:
						retry_after = response.msg["Retry-After"]
						self.console_message("Request limit exceeded, wait {limit} seconds before trying again".format(limit=retry_after))
						__ship(retry_attempts, int(retry_after)+1)
					else:
						self.console_message("Request limit exceeded")
				else:
					self.console_message("ship: failed on attempt {atmpt} (StatusCode: {sc}; Reason: {r})".format(sc=response.status, r=response.reason, atmpt=retry_attempts))
					raise Exception("ship exception")
			except Exception as ex:
				if (len(ex.args) > 0 and ex.args[0] == "PAYMENT_REQUIRED"):
					raise Exception("Either account is capped or an upgrade is required.")

				self.console_message("ship: exception shipping on attempt {atmpt}.".format(atmpt=retry_attempts))
				#self.console_message(ex, level=2)
				raise ex
				__ship(retry_attempts, 1)

		__ship(3)

	def set_bucket(self, bucket_name="", bucket_key="", retries=3):
		def __create_bucket(new_bucket_name, new_bucket_key, access_key):
			self.ship_to_api("/api/buckets", {'bucketKey': new_bucket_key, 'bucketName': new_bucket_name})

		if (bucket_key == None or bucket_key == ""):
			bucket_key = str(uuid.uuid4())

		self.BucketKey = bucket_key
		self.BucketName = bucket_name
		if (not self.Offline):
			if (self.Async):
				t = threading.Thread(target=__create_bucket, args=(bucket_name, bucket_key, self.AccessKey))
				t.daemon = False
				t.start()
			else:
				__create_bucket(bucket_name, bucket_key, self.AccessKey)
		else:
			self.console_message("Working in offline mode.", level=0)

	def console_message(self, message, level=1):
		if (self.DebugLevel >= level):
			print(message)

	def ship_messages(self, messages, retries=3):
		self.ship_to_api("/api/events", messages)
			

	def flush(self):
		if (self.Offline):
			self.console_message("flush: no need, in offline mode", level=2)
			return
		messages = []
		self.console_message("flush: checking queue", level=2)
		isEmpty = False
		while not isEmpty:
			try:
				m = self.LogQueue.popleft()
				messages.append(m)
			except IndexError:
				isEmpty = True
				self.console_message("flush: queue empty...", level=2)
		if len(messages) > 0:
			self.console_message("flush: queue not empty, shipping", level=2)
			
			self.ship_messages(messages)
			self.console_message("flush: finished flushing queue", level=2)


	def log_object(self, obj, key_prefix=None, epoch=None):
		if (epoch == None):
			epoch = time.time()

		if (key_prefix == None):
			key_prefix = "{}_".format(str(type(obj).__name__))
		elif (key_prefix != None and key_prefix != ""):
			key_prefix = "{}_".format(key_prefix)
		else:
			key_prefix = ""

		if (type(obj).__name__ == 'list'):
			i = 0
			for val in obj:
				key_name = "{}{}".format(key_prefix, i)
				self.log(key_name, val, epoch=epoch)
				i += 1
		elif (type(obj).__name__ == 'dict'):
			for key in obj:
				key_name = "{}{}".format(key_prefix, key)
				self.log(key_name, obj[key], epoch=epoch)
		else:
			for attr in dir(obj):
				if not isinstance(getattr(type(obj), attr, None), property):
					continue
				key_name = "{}{}".format(key_prefix, attr)
				self.log(key_name, getattr(obj, attr), epoch=epoch)


	def log(self, key, value, epoch=None):
		def __ship_buffer():
			i = self.BufferSize
			messages = []
			while(i > 0):
				try:
					m = self.LogQueue.popleft()
					messages.append(m)
				except IndexError:
					i = 0
					self.console_message("ship_buffer: queue empty")
				i = i - 1

			self.console_message("ship_buffer: shipping", level=2)
			self.ship_messages(messages)
			self.console_message("ship_buffer: finished shipping", level=2)

		timeStamp = time.time()
		gmtime = datetime.datetime.fromtimestamp(timeStamp)
		
		if epoch != None:
			try:
				gmtime = datetime.datetime.fromtimestamp(epoch)
				timeStamp = epoch
			except:
				self.console_message("epoch was overriden with invalid time, using current timstamp instead")
		
		formatted_gmTime = gmtime.strftime('%Y-%m-%d %H:%M:%S.%f')
		self.console_message("{time}: {key} {value}".format(key=key, value=value, time=formatted_gmTime))
		
		if (not self.Offline):
			if (len(self.LogQueue) >= self.BufferSize):
				self.console_message("log: queue size approximately at or greater than buffer size, shipping!", level=10)
				self.console_message("log: async is {async}".format(async=self.Async))
				if (self.Async):
					self.console_message("log: spawning ship thread", level=3)
					t = threading.Thread(target=__ship_buffer)
					t.daemon = False
					t.start()
				else:
					__ship_buffer()
		
			self.console_message("log: queueing log item", level=2)
			log_item = {
				"key": key,
				"value": value,
				"epoch": timeStamp
			}
			self.LogQueue.append(log_item)
		else:
			self.LocalFile.writerow([timeStamp, key, value])


	def close(self):
		self.IsClosed = True
		self.flush()
		if (self.MissedEvents != None):
			self.MissedEvents.close()
		if (self.Offline):
			self.console_message("closing local file handler", level=2)
			self.LocalFileHandler.close()

	def __del__(self):
		"""Try to close/flush the cache before destruction"""
		try:
			if (not self.IsClosed):
				self.close()
		except:
			if (self.DebugLevel >= 2):
				raise Exception("failed to close the buffer, make sure to explicitly call close() on the Streamer")
			else:
				self.console_message("failed to close the buffer, make sure to explicitly call close() on the Streamer", level=1)
